File: frontend/src/api/signals.ts
import { apiGet, apiPost } from "./client";

export type SignalSeverity =
  | "green"
  | "yellow"
  | "red"
  | "info"
  | "warning"
  | "critical"
  | "low"
  | "medium"
  | "high";

export interface Signal {
  id: string;
  type: string;
  severity: SignalSeverity;
  window: string;
  baseline_value: number | null;
  current_value: number | null;
  delta: number | null;
  explanation_seed: Record<string, unknown>;
  confidence: number;
}

export interface SignalsResponse {
  signals: Signal[];
  meta: Record<string, unknown>;
}

export async function fetchSignals(
  businessId: string,
  params: { start_date: string; end_date: string },
  signal?: AbortSignal
): Promise<SignalsResponse> {
  const query = new URLSearchParams({
    business_id: businessId,
    start_date: params.start_date,
    end_date: params.end_date,
  });
  return apiGet<SignalsResponse>(`/api/signals/v1?${query.toString()}`, { signal });
}

export type SignalStatus = "open" | "in_progress" | "resolved" | "ignored";

export type SignalState = {
  id: string;
  type: string | null;
  domain?: string | null;
  severity: SignalSeverity | null;
  status: SignalStatus;
  title: string | null;
  summary: string | null;
  updated_at: string | null;
};

export type SignalStateDetail = SignalState & {
  payload_json: Record<string, unknown> | null;
  fingerprint: string | null;
  detected_at: string | null;
  last_seen_at: string | null;
  resolved_at: string | null;
};

export type SignalStateResponse = {
  signals: SignalState[];
  meta: Record<string, unknown>;
};

export type SignalStatusUpdateInput = {
  status: SignalStatus;
  reason?: string | null;
  actor?: string | null;
};

export type SignalStatusUpdateResponse = {
  business_id: string;
  signal_id: string;
  status: SignalStatus;
  last_seen_at: string | null;
  resolved_at: string | null;
  resolution_note: string | null;
  reason: string | null;
  audit_id: string;
};

export function listSignalStates(businessId: string, signal?: AbortSignal) {
  const query = new URLSearchParams({ business_id: businessId });
  return apiGet<SignalStateResponse>(`/api/signals?${query.toString()}`, { signal });
}

export function getSignalDetail(
  businessId: string,
  signalId: string,
  signal?: AbortSignal
) {
  return apiGet<SignalStateDetail>(`/api/signals/${businessId}/${signalId}`, { signal });
}

export type SignalExplainEvidence = {
  key: string;
  label: string;
  value: string | number | boolean | null;
  unit?: string | null;
  as_of?: string | null;
  source: "ledger" | "state" | "derived" | "detector";
  anchors?: {
    txn_ids?: string[] | null;
    date_start?: string | null;
    date_end?: string | null;
    account_id?: string | null;
    vendor?: string | null;
    category?: string | null;
  } | null;
};


export type SignalExplainNextAction = {
  key: string;
  label: string;
  action: "acknowledge" | "snooze" | "resolve" | null;
  suggested_snooze_minutes?: number | null;
  requires_reason: boolean;
  rationale: string;
  guardrails?: string[] | null;
};


export type SignalExplainClearCondition = {
  summary: string;
  type: "threshold" | "trend" | "categorical";
  fields?: string[] | null;
  window_days?: number | null;
  comparator?: ">=" | "<=" | "==" | null;
  target?: number | string | null;
};

export type SignalExplainPlaybook = {
  id: string;
  title: string;
  description: string;
  kind: "inspect" | "adjust" | "decide";
  ui_target: "ledger" | "vendors" | "rules" | "categorize" | "assistant";
  deep_link?: string | null;
  requires_confirmation?: boolean | null;
};

export type SignalVerificationStatus = "met" | "not_met" | "unknown";

export type SignalExplainVerification = {
  status: SignalVerificationStatus;
  checked_at: string;
  facts: Array<{ key: string; label: string; value: string | number | boolean | null; source: "ledger" | "state" | "derived" | "detector" }>;
};

export type SignalExplainAudit = {
  id: string;
  event_type: string;
  actor: string | null;
  reason: string | null;
  status: string | null;
  created_at: string | null;
};

export type SignalExplainOut = {
  business_id: string;
  signal_id: string;
  state: {
    status: SignalStatus;
    severity: SignalSeverity | null;
    created_at: string | null;
    updated_at: string | null;
    last_seen_at: string | null;
    resolved_at: string | null;
    metadata: Record<string, unknown> | null;
    resolved_condition_met: boolean;
  };
  detector: {
    type: string;
    title: string;
    description: string;
    domain: string;
    default_severity: string | null;
    recommended_actions: Array<{
      action_id: string;
      label: string;
      rationale: string;
      parameters?: Record<string, unknown> | null;
    }>;
    evidence_schema: string[];
    scoring_profile: Record<string, unknown>;
  };
  evidence: SignalExplainEvidence[];
  related_audits: SignalExplainAudit[];
  next_actions: SignalExplainNextAction[];
  clear_condition: SignalExplainClearCondition | null;
  verification: SignalExplainVerification;
  playbooks: SignalExplainPlaybook[];
  links: string[];
};

export function getSignalExplain(
  businessId: string,
  signalId: string,
  signal?: AbortSignal
) {
  return apiGet<SignalExplainOut>(`/api/signals/${businessId}/${signalId}/explain`, { signal });
}

export function updateSignalStatus(
  businessId: string,
  signalId: string,
  payload: SignalStatusUpdateInput
): Promise<SignalStatusUpdateResponse>;
export function updateSignalStatus(
  businessId: string,
  signalId: string,
  payload: SignalStatusUpdateInput,
  options: { mode: "demo" }
): Promise<LegacySignalStatusUpdateResponse>;
export function updateSignalStatus(
  businessId: string,
  signalId: string,
  payload: SignalStatusUpdateInput,
  options?: { mode?: "demo" }
) {
  if (options?.mode === "demo") {
    return apiPost<LegacySignalStatusUpdateResponse>(
      `/demo/health/${businessId}/signals/${signalId}/status`,
      {
        status: payload.status,
        resolution_note: payload.reason ?? null,
      }
    );
  }
  return apiPost<SignalStatusUpdateResponse>(
    `/api/signals/${businessId}/${signalId}/status`,
    payload
  );
}

export type LegacySignalStatusUpdateResponse = {
  status: SignalStatus;
  resolved_at?: string | null;
  resolution_note?: string | null;
};

File: backend/app/api/routes/signals.py
from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Literal, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from backend.app.api.deps import get_current_user, require_membership_dep
from backend.app.db import get_db
from backend.app.models import User
from backend.app.services import signals_service
from backend.app.services.assistant_thread_service import append_receipt

router = APIRouter(prefix="/api/signals", tags=["signals"])


class SignalsResponse(BaseModel):
    signals: List[Dict[str, Any]]
    meta: Dict[str, Any]


class SignalStatusUpdateIn(BaseModel):
    status: str = Field(..., min_length=2, max_length=32)
    reason: Optional[str] = Field(default=None, max_length=500)
    actor: Optional[str] = Field(default=None, max_length=40)


class SignalStatusUpdateOut(BaseModel):
    business_id: str
    signal_id: str
    status: str
    last_seen_at: Optional[str]
    resolved_at: Optional[str]
    resolution_note: Optional[str]
    reason: Optional[str]
    audit_id: str


class SignalExplainStateOut(BaseModel):
    status: str
    severity: Optional[str]
    created_at: Optional[str]
    updated_at: Optional[str]
    last_seen_at: Optional[str]
    resolved_at: Optional[str]
    metadata: Optional[Dict[str, Any]]
    resolved_condition_met: bool = False


class SignalExplainClearConditionOut(BaseModel):
    summary: str
    type: Literal["threshold", "trend", "categorical"]
    fields: Optional[List[str]] = None
    window_days: Optional[int] = None
    comparator: Optional[Literal[">=", "<=", "=="]] = None
    target: Optional[Any] = None


class SignalExplainPlaybookOut(BaseModel):
    id: str
    title: str
    description: str
    kind: Literal["inspect", "adjust", "decide"]
    ui_target: Literal["ledger", "vendors", "rules", "categorize", "assistant"]
    deep_link: Optional[str] = None
    requires_confirmation: Optional[bool] = None


class SignalRecommendedActionOut(BaseModel):
    action_id: str
    label: str
    rationale: str
    parameters: Optional[Dict[str, Any]] = None


class SignalExplainDetectorOut(BaseModel):
    type: str
    title: str
    description: str
    domain: str
    default_severity: Optional[str]
    recommended_actions: List[SignalRecommendedActionOut]
    evidence_schema: List[str]
    scoring_profile: Dict[str, Any]


class LedgerAnchorQueryOut(BaseModel):
    source_event_ids: Optional[List[str]] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    accounts: Optional[List[str]] = None
    vendors: Optional[List[str]] = None
    categories: Optional[List[str]] = None
    search: Optional[str] = None
    direction: Optional[Literal["inflow", "outflow"]] = None


class SignalExplainEvidenceAnchorOut(LedgerAnchorQueryOut):
    pass


class SignalExplainEvidenceOut(BaseModel):
    key: str
    label: str
    value: Any
    unit: Optional[str] = None
    as_of: Optional[str] = None
    source: Literal["ledger", "state", "derived", "detector"]
    anchors: Optional[SignalExplainEvidenceAnchorOut] = None


class SignalExplainEvidenceSummaryOut(BaseModel):
    counts: Dict[str, Any] = Field(default_factory=dict)
    deltas: Dict[str, Any] = Field(default_factory=dict)
    rows: List[str] = Field(default_factory=list)


class SignalExplainExplanationOut(BaseModel):
    observation: str
    evidence: SignalExplainEvidenceSummaryOut
    implication: str


class SignalExplainLedgerAnchorOut(BaseModel):
    label: str
    query: LedgerAnchorQueryOut
    evidence_keys: List[str] = Field(default_factory=list)


class SignalExplainVerificationFactOut(BaseModel):
    key: str
    label: str
    value: Any
    source: Literal["ledger", "state", "derived", "detector"]


class SignalExplainVerificationOut(BaseModel):
    status: Literal["met", "not_met", "unknown"]
    checked_at: str
    facts: List[SignalExplainVerificationFactOut] = Field(default_factory=list)


class SignalExplainNextActionOut(BaseModel):
    key: str
    label: str
    action: Optional[Literal["acknowledge", "snooze", "resolve"]]
    suggested_snooze_minutes: Optional[int] = None
    requires_reason: bool
    rationale: str
    guardrails: Optional[List[str]] = None

class SignalExplainAuditOut(BaseModel):
    id: str
    event_type: str
    actor: Optional[str]
    reason: Optional[str]
    status: Optional[str]
    created_at: Optional[str]


class SignalExplainOut(BaseModel):
    business_id: str
    signal_id: str
    state: SignalExplainStateOut
    detector: SignalExplainDetectorOut
    evidence: List[SignalExplainEvidenceOut]
    related_audits: List[SignalExplainAuditOut]
    next_actions: List[SignalExplainNextActionOut]
    clear_condition: Optional[SignalExplainClearConditionOut] = None
    verification: SignalExplainVerificationOut
    playbooks: List[SignalExplainPlaybookOut] = Field(default_factory=list)
    explanation: SignalExplainExplanationOut
    ledger_anchors: List[SignalExplainLedgerAnchorOut] = Field(default_factory=list)
    links: List[str]


@router.get("", response_model=SignalsResponse, dependencies=[Depends(require_membership_dep())])
def list_signals(
    business_id: str = Query(...),
    db: Session = Depends(get_db),
):
    signals, meta = signals_service.list_signal_states(db, business_id)
    return SignalsResponse(signals=signals, meta=meta)


@router.get("/types", response_model=List[Dict[str, Any]])
def list_signal_types(
    include_inputs: Optional[bool] = Query(True),
):
    types = signals_service.available_signal_types()
    if not include_inputs:
        return [{"type": row["type"], "window_days": row["window_days"]} for row in types]
    return types


@router.get("/v1", response_model=SignalsResponse, dependencies=[Depends(require_membership_dep())])
def list_v1_signals(
    business_id: str = Query(...),
    start_date: date = Query(...),
    end_date: date = Query(...),
    db: Session = Depends(get_db),
):
    signals, meta = signals_service.fetch_signals(db, business_id, start_date, end_date)
    return SignalsResponse(signals=signals, meta=meta)


@router.post(
    "/{business_id}/{signal_id}/status",
    response_model=SignalStatusUpdateOut,
    dependencies=[Depends(require_membership_dep(min_role="staff"))],
)
def update_signal_status(
    business_id: str,
    signal_id: str,
    req: SignalStatusUpdateIn,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    result = signals_service.update_signal_status(
        db,
        business_id,
        signal_id,
        status=req.status,
        reason=req.reason,
        actor=req.actor or user.email,
    )
    append_receipt(
        db,
        business_id,
        {
            "action": "signal_status_updated",
            "actor": req.actor,
            "reason": req.reason,
            "signal_id": signal_id,
            "audit_id": result.get("audit_id"),
            "links": {
                "audit": f"/app/{business_id}/audit/{result.get('audit_id')}",
                "signal": f"/app/{business_id}/assistant?signalId={signal_id}",
            },
        },
        dedupe=False,
    )
    return result


@router.get("/{business_id}/{signal_id}", dependencies=[Depends(require_membership_dep())])
def get_signal_detail(
    business_id: str,
    signal_id: str,
    db: Session = Depends(get_db),
):
    return signals_service.get_signal_state_detail(db, business_id, signal_id)


@router.get(
    "/{business_id}/{signal_id}/explain",
    dependencies=[Depends(require_membership_dep())],
)
def get_signal_explain(
    business_id: str,
    signal_id: str,
    db: Session = Depends(get_db),
):
    return signals_service.get_signal_explain(db, business_id, signal_id)
